modal-js/src/sandbox.ts

```diff
@@ -1052,6 +1207,91 @@ export class ContainerProcess<R extends string | Uint8Array = any> {
   }
 }
 
+/** @ignore */
+export class ContainerProcessThroughRouter<
+  R extends string | Uint8Array = any,
+> {
+  stdin: ModalWriteStream<R>;
+  stdout: ModalReadStream<R>;
+  stderr: ModalReadStream<R>;
+  returncode: number | null = null;
+
+  readonly #taskId: string;
+  readonly #execId: string;
+  readonly #commandRouterClient: TaskCommandRouterClientImpl;
+  readonly #deadline: number | null;
+
+  constructor(
+    taskId: string,
+    execId: string,
+    commandRouterClient: TaskCommandRouterClientImpl,
+    params?: SandboxExecParams,
+    deadline?: number | null,
+  ) {
+    this.#taskId = taskId;
+    this.#execId = execId;
+    this.#commandRouterClient = commandRouterClient;
+    this.#deadline = deadline ?? null;
+
+    const mode = params?.mode ?? "text";
+    const stdout = params?.stdout ?? "pipe";
+    const stderr = params?.stderr ?? "pipe";
+
+    this.stdin = toModalWriteStream(
+      inputStreamRouter<R>(commandRouterClient, taskId, execId),
+    );
+
+    let stdoutStream = streamConsumingIter(
+      outputStreamRouter(
+        commandRouterClient,
+        taskId,
+        execId,
+        FileDescriptor.FILE_DESCRIPTOR_STDOUT,
+        this.#deadline,
+      ),
+    );
+    if (stdout === "ignore") {
+      stdoutStream.cancel();
+      stdoutStream = ReadableStream.from([]);
+    }
+
+    let stderrStream = streamConsumingIter(
+      outputStreamRouter(
+        commandRouterClient,
+        taskId,
+        execId,
+        FileDescriptor.FILE_DESCRIPTOR_STDERR,
+        this.#deadline,
+      ),
+    );
+    if (stderr === "ignore") {
+      stderrStream.cancel();
+      stderrStream = ReadableStream.from([]);
+    }
+
+    if (mode === "text") {
+      this.stdout = toModalReadStream(
+        stdoutStream.pipeThrough(new TextDecoderStream()),
+      ) as ModalReadStream<R>;
+      this.stderr = toModalReadStream(
+        stderrStream.pipeThrough(new TextDecoderStream()),
+      ) as ModalReadStream<R>;
+    } else {
+      this.stdout = toModalReadStream(stdoutStream) as ModalReadStream<R>;
+      this.stderr = toModalReadStream(stderrStream) as ModalReadStream<R>;
+    }
+  }
+
+  async wait(): Promise<number> {
```

Comment:
This logic differs from the python client, which sets the returncode member variable, and also has timeout handling logic

Reply:
Yes, this works differently, and I implemented it to mimic `ContainerProcess` in JS. Would suggest following up and changing this behavior in a separate follow-up, do you think that works?

Reply:
But could maybe get rid of the `returncode` member var? does that ever get set?